PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc

```diff
@@ -224,11 +233,33 @@ void pat::PATPackedCandidateProducer::produce(edm::Event& iEvent, const edm::Eve
           outPtrP->back().setAssociationQuality(pat::PackedCandidate::PVAssociationQuality(pat::PackedCandidate::UsedInFitTight));
         }
 	
-	if (puppiWeight.isValid()){
-	  reco::PFCandidateRef pkref( cands, ic );
-	  outPtrP->back().setPuppiWeight( (*puppiWeight)[pkref]);
-	  mappingPuppi[((*puppiCandsMap)[pkref]).key()]=ic;
-	}
+        if (puppiWeight.isValid()){
+           reco::PFCandidateRef pkref( cands, ic );
+                 // outPtrP->back().setPuppiWeight( (*puppiWeight)[pkref]);
+           
+           float puppiWeightVal = (*puppiWeight)[pkref];
+           float puppiWeightNoLepVal = 0.0;
+
+           // Check the "no lepton" puppi weights. 
+           // If present, then it is not a lepton, use stored weight
+           // If absent, it is a lepton, so set the weight to 1.0
+           if ( puppiWeightNoLep.isValid() ) {
+             // Look for the pointer inside the "no lepton" candidate collection.
+             auto foundNoLep = puppiCandsNoLep->end();
+             for ( auto s = puppiCandsNoLep->begin(),  s_end = puppiCandsNoLep->end(); s != s_end; ++s ) {       
+               if ( s->sourceCandidatePtr(0) == pkref->sourceCandidatePtr(0) ) {
```

Comment:
This line increases packedPFCandidates by 20 times.
Does this really need to go through the sourceCandidatePtr?
If needed, computation of sourceCandidatePtr(0) should be moved out of this nested loop and done outside of the loop over cands. A vector of Ptrs can then be compared in this nested loop. (x1000 speedup).
Also, this loop should be terminated after the first match.